http/src/v0/root_files/add.rs

```diff
@@ -1,111 +1,307 @@
+            UnsupportedField(name) => write!(fmt, "unsupported field name: {:?}", name),
+            UnsupportedContentType(t) => write!(fmt, "unsupported content-type: {:?} (supported: application/{{octet-stream,x-directory}})", t),
+            ResponseSerialization(e) => write!(fmt, "progress serialization failed: {}", e),
+            Persisting(e) => write!(fmt, "put_block failed: {}", e),
+            TreeGathering(g) => write!(fmt, "invalid directory tree: {}", g),
+            TreeBuilding(b) => write!(fmt, "constructed invalid directory tree: {}", b),
+        }
+    }
+}
 
-        let mut adder = FileAdder::default();
-        let mut total = 0u64;
+impl std::error::Error for AddError {}
+
+fn add_stream<St, E>(
+    ipfs: Ipfs<impl IpfsTypes>,
+    mut fields: MultipartStream<St, E>,
+    opts: AddArgs,
+) -> impl Stream<Item = Result<Bytes, AddError>> + Send + 'static
+where
+    St: Stream<Item = Result<Bytes, E>> + Send + Unpin + 'static,
+    E: Into<anyhow::Error> + Send + 'static,
+{
+    async_stream::try_stream! {
+
+        let mut tree_opts = TreeOptions::default();
+        if opts.wrap_with_directory {
+            tree_opts.wrap_with_directory();
+        }
 
-        loop {
-            let next = field
-                .try_next()
-                .await
-                .map_err(|e| StringError::from(format!("IO error: {}", e)))?;
+        let mut tree = BufferingTreeBuilder::new(tree_opts);
+        let mut buffer = BytesMut::new();
+
+        while let Some(mut field) = fields
+            .try_next()
+            .await?
+        {
+
+            let field_name = field.name().map_err(AddError::Header)?;
+            let filename = field.filename().map_err(AddError::Header)?;
+            let filename = percent_encoding::percent_decode_str(filename)
+                .decode_utf8()
+                .map_err(AddError::InvalidFilename)?;
+
+            let filename = if filename.starts_with('/') {
+                // normalize single first slash; seems similar to what js-ipfs does: filesystem
+                // test cases post with paths '/some-directory/...' and others post with
+                // 'some-directory/...'.
+
+                // since slash is single code point we can just
+                filename[1..].to_owned()
+            } else {
+                filename.into_owned()
+            };
+
+            let content_type = field.content_type().map_err(AddError::Header)?;
+
+            let next = match content_type {
+                "application/octet-stream" => {
+
+                    // files are file{,-1,-2,-3,..}
+                    let _ = if field_name != "file" && !field_name.starts_with("file-") {
+                        Err(AddError::UnsupportedField(field_name.to_string()))
+                    } else {
+                        Ok(())
+                    }?;
+
+                    let mut adder = FileAdder::default();
+                    // how much of bytes have we stored as blocks
```

Comment:
```suggestion
                    // how many bytes we have stored as blocks
```

Reply:
TBH I am quite surprised on how bad english I was able to write here :)